gateway/ln-gateway/src/lightning/ldk.rs

```diff
@@ -0,0 +1,502 @@
+    ) -> Result<GetRouteHintsResponse, LightningRpcError> {
+        // TODO: Return real route hints. Not strictly necessary but would be nice to
+        // have.
+        Ok(GetRouteHintsResponse {
+            route_hints: vec![],
+        })
+    }
+
+    async fn pay(
+        &self,
+        invoice: PayInvoiceRequest,
+    ) -> Result<PayInvoiceResponse, LightningRpcError> {
+        let parsed_invoice: Bolt11Invoice =
+            Bolt11Invoice::from_str(&invoice.invoice).map_err(|e| {
+                LightningRpcError::FailedPayment {
+                    failure_reason: format!("Failed to parse invoice: {e:?}"),
+                }
+            })?;
+
+        let payment_id = match self.node.bolt11_payment().send(&parsed_invoice) {
+            Ok(payment_id) => payment_id,
+            Err(e) => {
+                return Err(LightningRpcError::FailedPayment {
+                    failure_reason: format!("LDK payment failed to initialize: {e:?}"),
+                });
+            }
+        };
+
+        // TODO: Find a way to avoid looping/polling to know when a payment is
+        // completed. `ldk-node` provides `PaymentSuccessful` and `PaymentFailed`
+        // events, but interacting with the node event queue here isn't
+        // straightforward.
+        loop {
+            if let Some(payment_details) = self.node.payment(&payment_id) {
+                match payment_details.status {
+                    PaymentStatus::Pending => {}
+                    PaymentStatus::Succeeded => {
+                        if let PaymentKind::Bolt11 {
+                            preimage: Some(preimage),
+                            ..
+                        } = payment_details.kind
+                        {
+                            return Ok(PayInvoiceResponse {
+                                preimage: preimage.0.to_vec(),
+                            });
+                        }
+                    }
+                    PaymentStatus::Failed => {
+                        return Err(LightningRpcError::FailedPayment {
+                            failure_reason: "LDK payment failed".to_string(),
+                        });
+                    }
+                }
+            }
+            fedimint_core::runtime::sleep(Duration::from_millis(100)).await;
+        }
+    }
+
+    async fn route_htlcs<'a>(
+        mut self: Box<Self>,
+        _task_group: &TaskGroup,
+    ) -> Result<(RouteHtlcStream<'a>, Arc<dyn ILnRpcClient>), LightningRpcError> {
+        let route_htlc_stream = match self.htlc_stream_receiver_or.take() {
+            Some(stream) => Ok(Box::pin(ReceiverStream::new(stream))),
+            None => Err(LightningRpcError::FailedToRouteHtlcs {
+                failure_reason:
+                    "Stream does not exist. Likely was already taken by calling `route_htlcs()`."
+                        .to_string(),
+            }),
+        }?;
+
+        Ok((route_htlc_stream, Arc::new(*self)))
```

Comment:
This will move `self` rather than clone it, which is what we do in the other implementations. As far as I can tell this won't cause any issues (whereas with LND we need to clone it). I assume this was intentional?

Reply:
Yep! This was intentional. The reason that `ILnRpcClient::route_htlcs` consumes `self` and returns `Arc<Self>`  (IIUC) is so that the type checker can enforce that it cannot be called twice. That should still be enforced here.